### PR TITLE
PR: added Position and VacationRequest DTO's and manual mapping

### PR DESCRIPTION
**Pull request by nevence**

### CODE REVIEW

File: VacaYAY/BusinessLogicLayer/Dto/PositionDto/PositionForManipulationDto.cs [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using static DataAccesLayer.Entities.Enums;
+
+namespace BusinessLogicLayer.Dto.PositionDto
+{
+    public record PositionForManipulationDto
+    {
+        [Required(ErrorMessage = "This field is required.")]
```
Comment (pdavitkovic) [inline VacaYAY/BusinessLogicLayer/Dto/PositionDto/PositionForManipulationDto.cs]: Magic strings...

File: VacaYAY/BusinessLogicLayer/Dto/VacationRequestDto/VacationRequestForCreationDto.cs [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+using System.ComponentModel.DataAnnotations;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace BusinessLogicLayer.Dto.VacationRequestDto
+{
+    public record VacationRequestForCreationDto : VacationRequesForManipulationtDto
+    {
+        [Required(ErrorMessage = "This field is required.")]
```
Comment (pdavitkovic) [inline VacaYAY/BusinessLogicLayer/Dto/VacationRequestDto/VacationRequestForCreationDto.cs]: Same

File: VacaYAY/BusinessLogicLayer/Extensions/VacationRequestExtension.cs [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+namespace BusinessLogicLayer.Extensions
+{
+    public static class VacationRequestExtensions
+    {
+        public static VacationRequestDto MapToVacationRequestDto(this VacationRequest vacationRequest)
+        {
+            return new VacationRequestDto
+            {
+                EmployeeId = vacationRequest.EmployeeId,
+                FullName = string.Join(' ', vacationRequest.Employee.FirstName, vacationRequest.Employee.LastName).Trim(),
```
Comment (pdavitkovic) [inline VacaYAY/BusinessLogicLayer/Extensions/VacationRequestExtension.cs]: What about string interpolation here?

File: VacaYAY/BusinessLogicLayer/Services/PositionService.cs [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace BusinessLogicLayer.Services
+{
+    public sealed class PositionService : IPositionService
+    {
+
```
Comment (pdavitkovic) [inline VacaYAY/BusinessLogicLayer/Services/PositionService.cs]: Unnecessary white space

File: VacaYAY/BusinessLogicLayer/Services/PositionService.cs [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+            var positionEntity = positionForUpdate.MapToPositionCreation();
+            _repository.Position.Create(positionEntity);
+            await _repository.SaveAsync();
+
+            return positionEntity.Id;
+        }
+
+        public async Task DeletePositionAsync(int id)
+        {
+            var positionEntity = await _repository.Position.FindByCondition(v => v.Id.Equals(id), false).SingleOrDefaultAsync();
```
Comment (pdavitkovic) [inline VacaYAY/BusinessLogicLayer/Services/PositionService.cs]: SingleOrDefaultAsync also has predicate overload. Think that find by condition is unecessery

File: VacaYAY/BusinessLogicLayer/Services/PositionService.cs [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+            Guard.ThrowIfNotFound(positionEntity, id);
+
+            _repository.Position.Delete(positionEntity);
+            await _repository.SaveAsync();
+        }
+
+        public async Task<PositionDto> GetPositionAsync(int id)
+        {
+            var positionEntity = await _repository.Position.FindByCondition(v => v.Id.Equals(id), false).SingleOrDefaultAsync();
```
Comment (pdavitkovic) [inline VacaYAY/BusinessLogicLayer/Services/PositionService.cs:45]: Same

File: VacaYAY/BusinessLogicLayer/Services/PositionService.cs [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+        {
+            var positionsEntity = await _repository.Position.FindAll(false).ToListAsync();
+
+            var positionsDto = positionsEntity.MapToPositionsDto();
+            return positionsDto;
+        }
+
+        public async Task UpdatePositionAsync(int id, PositionForUpdateDto positionForUpdate)
+        {
+            var positionEntity = await _repository.Position.FindByCondition(v => v.Id.Equals(id), true).SingleOrDefaultAsync();
```
Comment (pdavitkovic) [inline VacaYAY/BusinessLogicLayer/Services/PositionService.cs:62]: Same

File: VacaYAY/BusinessLogicLayer/Services/VacationRequestService.cs [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+            var vacationRequestEntity = vacationRequestForCreation.MapToVacationRequestCreation();
+            _repository.VacationRequest.Create(vacationRequestEntity);
+            await _repository.SaveAsync();
+            
+            return vacationRequestEntity.Id;
+        }
+
+        public async Task DeleteVacationRequestAsync(int id)
+        {
+            var vacationRequestEntity = await _repository.VacationRequest.FindByCondition(v => v.Id.Equals(id), false).SingleOrDefaultAsync();
```
Comment (pdavitkovic) [inline VacaYAY/BusinessLogicLayer/Services/VacationRequestService.cs]: Same

File: VacaYAY/BusinessLogicLayer/Services/VacationRequestService.cs [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+            Guard.ThrowIfNotFound(vacationRequestEntity, id);
+
+            _repository.VacationRequest.Delete(vacationRequestEntity);
+            await _repository.SaveAsync();
+        }
+
+        public async Task<VacationRequestDto> GetVacationRequestAsync(int id)
+        {
+           var vacationRequestEntity = await _repository.VacationRequest.FindByCondition(v => v.Id.Equals(id), false).SingleOrDefaultAsync();
```
Comment (pdavitkovic) [inline VacaYAY/BusinessLogicLayer/Services/VacationRequestService.cs]: Same

File: VacaYAY/BusinessLogicLayer/Services/VacationRequestService.cs [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+            var vacationRequestsEntity = await _repository.VacationRequest.FindByCondition(v => v.EmployeeId.Equals(employeeId), false).ToListAsync();
+
+            var vacationRequestsDto = vacationRequestsEntity.MapToVacationRequestsDto();
+            return vacationRequestsDto;
+
+        }
+
+        public async Task UpdateVacationRequestAsync(int id, VacationRequestForUpdateDto vacationRequestForUpdate)
+        {
+            var vacationRequestEntity = await _repository.VacationRequest.FindByCondition(v => v.Id.Equals(id), true).SingleOrDefaultAsync();
```
Comment (pdavitkovic) [inline VacaYAY/BusinessLogicLayer/Services/VacationRequestService.cs:72]: Same

**Review by pdavitkovic — CHANGES_REQUESTED** [review]

File: VacaYAY/BusinessLogicLayer/Extensions/VacationRequestExtension.cs [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+namespace BusinessLogicLayer.Extensions
+{
+    public static class VacationRequestExtensions
+    {
+        public static VacationRequestDto MapToVacationRequestDto(this VacationRequest vacationRequest)
+        {
+            return new VacationRequestDto
+            {
+                EmployeeId = vacationRequest.EmployeeId,
+                FullName = $"{vacationRequest.Employee.FirstName} {vacationRequest.Employee.LastName}".Trim(),
```
Comment (pdavitkovic) [inline VacaYAY/BusinessLogicLayer/Extensions/VacationRequestExtension.cs]: It Trim() necessery?

File: VacaYAY/BusinessLogicLayer/Services/PositionService.cs [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+            var positionEntity = positionForUpdate.MapToPositionCreation();
+            _repository.Position.Create(positionEntity);
+            await _repository.SaveAsync();
+
+            return positionEntity.Id;
+        }
+
+        public async Task DeletePositionAsync(int id)
+        {
+            var positionEntity = await _repository.Position.FindAll(true).SingleOrDefaultAsync(v => v.Id.Equals(id));
```
Comment (pdavitkovic) [inline VacaYAY/BusinessLogicLayer/Services/PositionService.cs]: Why did you use here FindAll() + SingleOrDefaultAsync()?

File: VacaYAY/BusinessLogicLayer/Services/PositionService.cs [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+            Guard.ThrowIfNotFound(positionEntity, id);
+
+            _repository.Position.Delete(positionEntity);
+            await _repository.SaveAsync();
+        }
+
+        public async Task<PositionDto> GetPositionAsync(int id)
+        {
+            var positionEntity = await _repository.Position.FindAll(false).SingleOrDefaultAsync(v => v.Id.Equals(id));
```
Comment (pdavitkovic) [inline VacaYAY/BusinessLogicLayer/Services/PositionService.cs]: Same here

File: VacaYAY/BusinessLogicLayer/Services/VacationRequestService.cs [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+            var vacationRequestEntity = vacationRequestForCreation.MapToVacationRequestCreation();
+            _repository.VacationRequest.Create(vacationRequestEntity);
+            await _repository.SaveAsync();
+            
+            return vacationRequestEntity.Id;
+        }
+
+        public async Task DeleteVacationRequestAsync(int id)
+        {
+            var vacationRequestEntity = await _repository.VacationRequest.FindAll(true).SingleOrDefaultAsync(v => v.Id.Equals(id));
```
Comment (pdavitkovic) [inline VacaYAY/BusinessLogicLayer/Services/VacationRequestService.cs]: Same

File: VacaYAY/BusinessLogicLayer/Services/VacationRequestService.cs [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+            Guard.ThrowIfNotFound(vacationRequestEntity, id);
+
+            _repository.VacationRequest.Delete(vacationRequestEntity);
+            await _repository.SaveAsync();
+        }
+
+        public async Task<VacationRequestDto> GetVacationRequestAsync(int id)
+        {
+            var vacationRequestEntity = await _repository.VacationRequest.FindAll(false).SingleOrDefaultAsync(v => v.Id.Equals(id));
```
Comment (pdavitkovic) [inline VacaYAY/BusinessLogicLayer/Services/VacationRequestService.cs]: Same

File: VacaYAY/BusinessLogicLayer/Services/VacationRequestService.cs [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+            var vacationRequestsEntity = await _repository.VacationRequest.FindByCondition(v => v.EmployeeId.Equals(employeeId), false).ToListAsync();
+
+            var vacationRequestsDto = vacationRequestsEntity.MapToVacationRequestsDto();
+            return vacationRequestsDto;
+
+        }
+
+        public async Task UpdateVacationRequestAsync(int id, VacationRequestForUpdateDto vacationRequestForUpdate)
+        {
+            var vacationRequestEntity = await _repository.VacationRequest.FindAll(true).SingleOrDefaultAsync(v => v.Id.Equals(id));
```
Comment (pdavitkovic) [inline VacaYAY/BusinessLogicLayer/Services/VacationRequestService.cs]: Same

**Review by pdavitkovic — CHANGES_REQUESTED** [review]

File: VacaYAY/BusinessLogicLayer/Extensions/VacationRequestExtension.cs (hunk 11 above)
Comment (nevence) [reply VacaYAY/BusinessLogicLayer/Extensions/VacationRequestExtension.cs]: i thought about a case where user could accidentally type in a blank space at the end in FirstName or LastName when creating an Employee, so i thought Trim would handle this situation, however you are right, it makes more sense for me to add Trim when creating an Employee, or any entity, not in here. what do you think, should I add Trim when creating and adding new entites in DB or is it not necessary?

File: VacaYAY/BusinessLogicLayer/Services/PositionService.cs (hunk 12 above)
Comment (nevence) [reply VacaYAY/BusinessLogicLayer/Services/PositionService.cs]: public IQueryable<T> FindAll(bool trackChanges)
            {
                if (!trackChanges)
                    return ApplicationDbContext.Set<T>().AsNoTracking();
                else
                    return ApplicationDbContext.Set<T>();
            }

this is my FindAll method 

i can use singleordeafultasync only on IQueryable types. Visual Studio doesn't allow me to directly use AsNoTracking().SingleOrDefaultAsync on _repository.VacationRequest:

(_repository.VacationRequest.AsNoTracking().SingleOrDefaultAsync( v=> v.Id.Equals(id)); )
(

File: VacaYAY/BusinessLogicLayer/Extensions/VacationRequestExtension.cs (hunk 11 above)
Comment (pdavitkovic) [reply VacaYAY/BusinessLogicLayer/Extensions/VacationRequestExtension.cs]: Not necessery

File: VacaYAY/BusinessLogicLayer/Services/PositionService.cs (hunk 12 above)
Comment (pdavitkovic) [reply VacaYAY/BusinessLogicLayer/Services/PositionService.cs]: You can execute SingleOrDefaultAsync directly on Position

**Review by pdavitkovic — CHANGES_REQUESTED** [review]

**Pull request merged**